# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv('../data/supermarket_sales.csv')

In [3]:
type(df)

pandas.core.frame.DataFrame

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
df.sort_values('Unit price', ascending=False).head()

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [16]:
df[df["Product line"] == "Electronic accessories"].sort_values(['Quantity', 'Unit price'], ascending=[False, False]).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

In [21]:
df.query('`Product line` == "Electronic accessories"').nlargest(1, ['Quantity', 'Unit price'])

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [37]:
 def category_mapping (value):
     if value  >= 8:
         return 'High'
     elif value  >= 6:
         return 'Medium'
     else:
         return 'Low'

In [51]:
df['Rating Category'] = df.Rating.apply(category_mapping)
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10

In [29]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating', 'Rating Category'],
      dtype='object')

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [40]:
df['Rating Category'].value_counts(normalize=True) * 100

Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [41]:
time = '11:28'

In [43]:
def get_hour (t):
    return (int(t[:2]))

In [46]:
print(get_hour(time))

11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [47]:
df['Hour'] = df.Time.apply(get_hour)

In [50]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [54]:
df.Hour.value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

Найбільше покупок відбувається о 19:00, 13:00–15:00 та 10:00.
Це логічно — люди купують після роботи, під час обідньої перерви, та перед роботою коли мають вільний час.
Видно, що магазін працює з 10:00 до 20:00

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [58]:
round(df.groupby('City').Total.mean(),4)

City
Mandalay     319.8725
Naypyitaw    337.0997
Yangon       312.3540
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [68]:
df.groupby('City').Total.agg(['count', 'mean', 'min', 'max']).round(2)

count    mean    min      max
City                                    
Mandalay     332  319.87  18.64  1022.49
Naypyitaw    328  337.10  10.68  1042.65
Yangon       340  312.35  12.69  1039.29

1. Набільша кількість покупок в **Yangon (340)** - свідчить про більшу кільксть крієнтів або їх вищу активність.
2. Найвищийй середній чек в **Naypyitaw (337)**  - може вкасувати на вищу купівельну спроможність, або представлення більш дорожчого товару. Також в цьому місті наявний найвищий чек - **1042**
3. **Mandalay** має найнижчий максимальний чек **1022**, але найвищий мінімальний -  **18.6**, що може вказувати на більш рівномірні покупки.
4. В цілому **різниця між мінімальними та максимальними чеками** у всіх містах доволі значна, що вказує на різну купівельну спроможність клієнтів або широкий асортимент.

**Ремомендації :**  
1. Naypyitaw.
   - В цьому місті спостерігається менша кількість покупок, але найвищий середній час, отже аудиторія має вищу купівельну спроможність, тому варто посилити маркетинг для залучення нових клієнтів та утримання наявних (система лояльності)
2. Yangon
   - Спостерігається високий попит, але середній чек нижчий (можливо клієнти з нижчою купівельною спроможністю). отже варто стимулювати додаткові покупки (акції, знижки, програми лояльності).
3. Mandalay:
   - спостерігаємо середнє значення між кількістю покупок і чеком,отже рекомендовано утримувати клієнтів за допомогою персоналізованих пропозицій і стабільної якості сервісу.
4. Загалом:
   - Провести ціновий аналіз і сегментацію клієнтів для розробки індивідуальних стратегій по містах.
   - Перевірити асортимент — можливо, є потенціал для збільшення середнього чеку через преміум-продукти.
   - Перевірити асортисент та популярність товарів.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [70]:
df.to_csv('supermarket_sales_enriched.csv', index=False)

In [71]:
pd.read_csv('supermarket_sales_enriched.csv', nrows = 5)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10